libff/algebra/fields/field.hpp

```diff
@@ -0,0 +1,90 @@
+/** @file
+ *****************************************************************************
+ Declaration of common API for all finite fields.
+
+ Currently NOT used by the fields in this library. This class is not actually
+ the parent class of any field. All APIs are enforced through tests instead.
+
+ The reason for this is to ensure high performance of all fields. This class
+ exists as documentation for common API between fields.
+
+ Includes two types of fields, F[p^n] for selected n and F[2^n] for a separate
+ range of n. All of these finite fields must implement all functions declared
+ in this class.
+ *****************************************************************************
+ * @author     This file is part of libff, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+#include <libff/algebra/field_utils/bigint.hpp>
+#include <vector>
+
+namespace libff {
+
+template<typename T>
+class Field;
+
+/* The type parameter T is intended to be set to the child class
+ * when this class is extended. For example,
+ * class Fp_model : public Field<Fp_model> ...
+ */
+template<typename T>
+class Field {
+public:
+#ifdef PROFILE_OP_COUNTS // NOTE: op counts are affected when you exponentiate with ^
+    static long long add_cnt;
+    static long long sub_cnt;
+    static long long mul_cnt;
+    static long long sqr_cnt;
+    static long long inv_cnt;
+#endif
+
+    virtual T& operator+=(const T& other) = 0;
+    virtual T& operator-=(const T& other) = 0;
+    virtual T& operator*=(const T& other) = 0;
+    virtual T& operator^=(const unsigned long pow) = 0;
+    template<mp_size_t m>
+    virtual T& operator^=(const bigint<m> &pow) = 0;
+
+    virtual T& square() = 0;
+    virtual T& invert() = 0;
+
+    virtual T operator+(const T& other) const;
+    virtual T operator-(const T& other) const;
+    virtual T operator*(const T& other) const;
+    virtual T operator-() const = 0;
+
+    virtual T squared() const;
+    virtual T inverse() const;
+    /** HAS TO BE A SQUARE (else does not terminate). */
+    virtual T sqrt() const = 0;
+
+    virtual T operator^(const unsigned long pow) const;
+    template<mp_size_t m>
+    virtual T operator^(const bigint<m> &pow) const;
+
+    bool operator==(const T& other) const = 0;
+    bool operator!=(const T& other) const = 0;
+    bool is_zero() const = 0;
+
+    void print() const = 0;
+
+    void randomize() = 0;
+    void clear() = 0;
+
+    // the following should be defined in child classes, but are static so they can't be inherited
+    static T zero();
+    static T one();
+    static T random_element();
+    /** Equals 1 for prime field Fp. */
+    static constexpr std::size_t extension_degree();
+    static std::size_t size_in_bits();
```

Comment:
If a modulus is `2^n + 1` (or any n+1 bit number), does this return `n` or `n+1`?

Reply:
IIRC, `size_in_bits(x)` should return `ceil(log_2(x))` (i.e. `n+1` in your example above) (`capacity(x)`, however, returns `floor(log_2(x))` (i.e. `n` in your eg))

Reply:
That makes sense. Should we instead call them `floor_size_in_bits` and `ceil_size_in_bits` to remove ambiguity?

Reply:
I think adding a comment explaining should be fine?

Reply:
hrmm, I'd strongly prefer an unambiguous function name. This feels like it can cause confusion / off-by-one soundness errors to downstream users.

Reply:
Right now only Fp implements capacity(), I'll change it to make all fields implement it, renamed floor_size_in_bits()